File: apps/processor-node/src/storage.ts
// apps/processor-node/src/storage.ts
import os from 'os';
import path from 'path';
import fs from 'fs';
import { Storage } from '@google-cloud/storage';
import { Firestore, FieldValue } from '@google-cloud/firestore';
import Logger from "./utils/logger";

const bucketName = process.env.GCS_BUCKET!;
const storage = new Storage();
const bucket = storage.bucket(bucketName);
const firestore = new Firestore();

type TJobStatus = 'processing'
    | 'done'
    | 'error'
    | 'received_processor'
    | 'preparing_data'
    | 'before_processing'
    | 'uploading'

async function downloadFile(src: string, dest: string): Promise<void> {
    await new Promise<void>((resolve, reject) => {
        const readStream = bucket.file(src).createReadStream();
        const writeStream = fs.createWriteStream(dest);
        readStream
            .on('error', reject)
            .pipe(writeStream)
            .on('error', reject)
            .on('finish', resolve)
    });
}

/** Download audio and cover into tmp directory */
export async function downloadSources(jobId: string): Promise<{ audioPath: string; coverPath: string }> {
    const tmpDir = os.tmpdir();
    const audioPath = path.join(tmpDir, `${jobId}-audio.mp3`);
    const coverPath = path.join(tmpDir, `${jobId}-cover.jpeg`);

    await Promise.all([
        downloadFile(`uploads/${jobId}/audio.mp3`, audioPath),
        downloadFile(`uploads/${jobId}/cover.jpeg`, coverPath)
    ]);

    return { audioPath, coverPath };
}

/** Upload the rendered video and return its public URL */
export async function uploadResult(jobId: string, outputPath: string): Promise<string> {
    const destination = `outputs/${jobId}/video.mp4`;
    await bucket.upload(outputPath, { destination, contentType: 'video/mp4' });
    const file = bucket.file(destination);
    await file.makePublic();
    return `https://storage.googleapis.com/${bucketName}/${destination}`;
}

/** Update job status in Firestore */
export async function updateJobStatus(jobId: string, status: TJobStatus, videoUrl?: string): Promise<void> {
    const docRef = firestore.collection('jobs').doc(jobId);
    const data: any = { status, updatedAt: FieldValue.serverTimestamp() };

    if (videoUrl) data.videoURL = videoUrl;

    Logger.debug('updateJobStatus : jobId=', jobId, 'status=', status);

    await docRef.update(data);
}

/** Get job settings from Firestore */
export async function getJobSettings(jobId: string): Promise<any> {
    const doc = await firestore.collection('jobs').doc(jobId).get();
    return doc.data()?.settings;
}

export async function updateJobProgress(jobId: string, progress: number) {
    await firestore
        .doc(`jobs/${jobId}`)
        .set({ progress, updatedAt: FieldValue.serverTimestamp() }, { merge: true });
}

File: apps/processor/src/storage.py
import os
import tempfile
from google.cloud import storage, firestore

# Инициализация клиента (в app.py переменные окружения должны быть выставлены)
bucket_name = os.getenv("GCS_BUCKET")
storage_client = storage.Client()
bucket = storage_client.bucket(bucket_name)
firestore_client = firestore.Client()


def download_sources(job_id: str, tmp_dir: str) -> (str, str):
    """
    Скачивает audio.mp3 и cover.jpg в папку tmp_dir, возвращает пути.
    """
    audio_blob = bucket.blob(f"uploads/{job_id}/audio.mp3")
    cover_blob = bucket.blob(f"uploads/{job_id}/cover.jpg")
    audio_path = os.path.join(tmp_dir, "audio.mp3")
    cover_path = os.path.join(tmp_dir, "cover.jpg")
    audio_blob.download_to_filename(audio_path)
    cover_blob.download_to_filename(cover_path)
    return audio_path, cover_path


def upload_result(job_id: str, output_path: str) -> str:
    """
    Загружает видео, делает его публичным (или можно генерить signed URL) и возвращает URL.
    """
    dest_blob = bucket.blob(f"outputs/{job_id}/video.mp4")
    dest_blob.upload_from_filename(output_path, content_type="video/mp4")
    dest_blob.make_public()
    return dest_blob.public_url


def update_job_status(job_id: str, status: str, video_url: str = None):
    """
    Обновляет документ jobs/{job_id} в Firestore.
    """
    doc_ref = firestore_client.collection("jobs").document(job_id)
    data = {"status": status}
    if video_url:
        data["videoURL"] = video_url
    doc_ref.update(data)